### ISSUE TITLE
Using Delay and await can result in 100% CPU

### ISSUE THREAD

**Issue by LokiMidgard** [theraot/Theraot#20]
I tried to figure how ```Task.Delay()``` will perform compared to the netFramework version and alternativly the use of ```Thread.Sleep()```.

I started 1000 tasks and limeted the ThreadPool to 10 Threads (see code below). I didn't expected it to work like the original tasks with await, because I asume that needs support of native Framework capabilitys. 

But it actually performed far worse then using Sleeps. Thats supriesd me. But I didn't understand why. I didn't managed to understand how everything works with GetAwaiter and the other functions.

On my testmchine it didn't print more then the first 10 WriteLines before using 100% on all CPUs. After 5 to 10 minutes I killed the process without more output.

```c#

#define DELAY
using System;
using System.Linq;
using System.Threading.Tasks;

namespace ClassLibrary1
{
    public class Class1
    {
        public async static Task Run()
        {
            System.Threading.ThreadPool.SetMaxThreads(10, 10);
            var tasks = Enumerable.Range(0, 1000).Select(x =>
              Task.Run(async () =>
              {
                  var guid = Guid.NewGuid();
                  var id1 = System.Threading.Thread.CurrentThread.ManagedThreadId;
                  Console.WriteLine($"enter {id1}");
#if DELAY
                  await Task.Delay(5 * 1000);
#else
                  System.Threading.Thread.Sleep(5 * 1000);
#endif
                  var id2 = System.Threading.Thread.CurrentThread.ManagedThreadId;
#if DELAY
                  await Task.Delay(5 * 1000);
#else
                  System.Threading.Thread.Sleep(5 * 1000);
#endif
                  var id3 = System.Threading.Thread.CurrentThread.ManagedThreadId;
                  var result = (guid: guid, id1: id1, id2: id2, id3: id3);
                  Console.WriteLine(result);
                  return result;
              })).ToArray();
            var allTask = await Task.WhenAll(tasks);
        }

    }
}


```

**Comment by theraot** [comment]
The shor explanation of how Task.Delay is implemented by creating a Timeout

https://github.com/theraot/Theraot/blob/master/Core/System/Threading/Tasks/Task.delay.net35.cs#L127

Which wraps a Timer

https://github.com/theraot/Theraot/blob/master/Core/Theraot/Threading/Timeout.cs#L238

On the callback it would set the task as completed. As such, it is not
spinning. However I admit I should figure out a way to pool timers to
reduce memory footprint.

There is a bit of extra work if you check the status of the task, in which
case it will look for the remaining time and if it has completed it will
stop the timer and call the callback synchronously.

---

I haven't tested Task.Delay or async/wait reducing the size of the
ThreadPool, do you get the same result if you do not do that? The Timer
will run the callback in a worker thread, so I suspect that is part of the
problem.

---

As for await, it requires support of the compiler but not the runtime. The
compiler will convert the code into continuations. I guess I should review
that code once more.




2017-12-21 11:41 GMT-05:00 Patrick Kranz <notifications@github.com>:

**Comment by LokiMidgard** [comment]
Thanks for the fast response.

I think there is some other issue in my configuration. Just using one Delay raises my cpu from 5-10% to 30% CPU. Using ```TaskCompletionSource<object>``` has the same Problem. Using 10 at once ends in 100% CPU.

In that situation all working threads are looping ```Task.PrivateWait(...)```. This is the stacktrace:


```
>	Theraot.Core.dll!System.Threading.Tasks.Task.PrivateWait(System.Threading.CancellationToken cancellationToken = IsCancellationRequested = false, bool throwIfExceptional = true) Line 692	C#	Symbols loaded.
 	Theraot.Core.dll!System.Threading.Tasks.Task.Wait(System.Threading.CancellationToken cancellationToken = IsCancellationRequested = false) Line 377	C#	Symbols loaded.
 	Theraot.Core.dll!System.Threading.Tasks.Task.Wait() Line 372	C#	Symbols loaded.
 	Theraot.Core.dll!System.Threading.Tasks.Task<(System.Guid, int, int)>.Result.get() Line 62	C#	Symbols loaded.
 	Theraot.Core.dll!System.Threading.Tasks.Task.Run.AnonymousMethod__0() Line 108	C#	Symbols loaded.
 	Theraot.Core.dll!System.Threading.Tasks.Task<(System.Guid, int, int)>.InnerInvoke() Line 82	C#	Symbols loaded.
 	Theraot.Core.dll!System.Threading.Tasks.Task.Execute() Line 489	C#	Symbols loaded.
 	Theraot.Core.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal() Line 520	C#	Symbols loaded.
 	Theraot.Core.dll!System.Threading.Tasks.Task.ExecuteEntry(bool preventDoubleExecution = true) Line 487	C#	Symbols loaded.
 	Theraot.Core.dll!System.Threading.Tasks.ThreadPoolTaskScheduler.TaskExecuteCallback(object obj = {System.Threading.Tasks.Task<(System.Guid, int, int)>}) Line 130	C#	Symbols loaded.
 	[External Code]		Annotated Frame

```

For test purposes I added a ```Thread.Sleep(100)``` at the end of the loop. Which reduced the CPU down to almost nothing.

```c#
   private void PrivateWait(CancellationToken cancellationToken, bool throwIfExceptional)
        {
            var done = false;
            while (!done)
            {
                CancellationCheck(cancellationToken);
                switch (Status)
                {
                    case TaskStatus.WaitingToRun:
                        WaitAntecedent(CancellationToken);
                        ExecutingTaskScheduler.InernalTryExecuteTaskInline(this, true);
                        break;

                    case TaskStatus.Created:
                    case TaskStatus.WaitingForActivation:
                    case TaskStatus.Running:
                    case TaskStatus.WaitingForChildrenToComplete:
                        var waitHandle = _waitHandle.Value;
                        if (_waitHandle.IsAlive)
                        {
                            waitHandle.Wait(cancellationToken);
                        }
                        break;

                    case TaskStatus.RanToCompletion:
                    case TaskStatus.Canceled:
                        done = true;
                        break;

                    case TaskStatus.Faulted:
                        if (throwIfExceptional)
                        {
                            ThrowIfExceptional(true);
                        }
                        done = true;
                        break;
                }
                System.Threading.Thread.Sleep(100);
            }
#if DEBUG
            if (!IsCompleted)
            {
                Diagnostics.Debugger.Break();
            }
#endif
        }
```

**Comment by theraot** [comment]
That is strange. Makes me thing it is a case of code reordering.

The code is not hard to follow, it is a state machine: if the task is yet
to be executed it will try to run it inline which should put it to
completed or faulted. Otherwise, it will try to use wait handle. The only
case where the wait handle would not be alive is when the task is disposed,
but that can only happen if it ran to completion.

I would expect it to do a few cycles because it may read an state but the
task changed state before the appropiate operation was done. For example,
if it find the task is yet to run, but before it tries to run it inline it
starts, of if its running but before it gets to use the wait handle the
task completes and gets disposed. Yet, it any case, I would expect it to do
only a few iterations.

---

Status is a volatile read, and that would prevent it to optimize it away.
But the read of the wait handle is not. The compiler could move read of the
handle inside the if, which is not correct[1], but should not be a problem
if the if gives true... and again, the wait handle is created on the
constructor, and should stay there until the task is disposed, which would
only be possible in a complete state.

[1]: Using `if (!ReferenceEquals(waitHandle, null))` like in
https://github.com/theraot/Theraot/blob/v1/Core/System/Threading/Tasks/Task.net35.cs#L624
(which is the dispose code) would be better - but I would not expect that
to fix this problem.

---

As mitigation, we could use `var spinWait = new SpinWait();` before the
loop, and `spinWait.SpinOnce();` in the loop. See
https://github.com/theraot/Theraot/blob/14341ceac33ddf48265e6cb409f081f0a889ec15/Core/System/Threading/Tasks/Task.net35.cs#L751
for an example. Not really a fix. I would need to figure out why the code
is stuck there.

---

That is all I can think of right now. Will come back to this later.





2017-12-22 2:56 GMT-05:00 Patrick Kranz <notifications@github.com>:

**Comment by LokiMidgard** [comment]
I'm curious, if it would work as intended, would it block the thread on the await, or would the thread start working on another task like .NetFramwork async would do?

**Comment by theraot** [comment]
It should block the awaiting thread. As I said on my first replay, it
requires support from the compiler to work, the compiler will rewrite the
code into task completion source and continuations. Thus, if that and other
related APIs are available and working, it should work.

I did go over how async/await work in more detail over stackoverflow:
https://stackoverflow.com/a/47492965/402022 (the original answer skims over
it, the extended one give you the details). I also did a demo of equivalent
code with and withou async/await by doing something similar to what the
compiler does:

- With: https://dotnetfiddle.net/cjlrqS
- Without: https://dotnetfiddle.net/fgF3Wu

The performance in the backport will not be the same. However you should
not have a livelock, the code should be getting stuck there.


2017-12-22 3:59 GMT-05:00 Patrick Kranz <notifications@github.com>:

**Comment by theraot** [comment]
I am back!

I have been running the code you provided, and doing so I have reached a
situation where a task status is in WaitingForActivation but the wait
handle for the task is null, given that the wait handle is unaviable the
code is spin waiting. I will be working on a fix.

2017-12-22 4:28 GMT-05:00 Alfonso J. Ramos <theraot@gmail.com>:

**Comment by theraot** [comment]
There is a working fix in the repository, please try it out.

I will upload the nuget when I have a chance.

2017-12-27 19:53 GMT-05:00 Alfonso J. Ramos <theraot@gmail.com>:

**Comment by theraot** [comment]
Scratch that, I got Heisenbugs.

2017-12-28 3:58 GMT-05:00 Alfonso J. Ramos <theraot@gmail.com>:

**Comment by LokiMidgard** [comment]
Ok,

I'm not sure if I can check this the next weeks. I'm on vacation. But good to know it's not some wired configuration that just happens on my machine.

**Comment by theraot** [comment]
@LokiMidgard 

I know this took long... I had been busy with other - not open - projects.

I did a global code review and fixed all the related problems I found, and some unrelated too. I pushed a new nuget (just finished uploading as I type).

I also got rid of some tests that were based on race conditions (which means they will fail unpredictably). However, I cannot get rid of race condition in tests, because a test running indefinitely is not an useful fail condition.

Anyway, let me know if the libraries give you anymore trouble.
